### PR TITLE
Fix and tweak parrot behaviour

### LINKED ISSUE #13640: Poly / parrots can teleport through glass
**Issue Description**:

Poly / parrots can perch on consoles on the other side of glass windows, teleporting them through glass.

**What did you expect to happen**: 

Poly to be unable to perch like that

**What happened instead**:

Poly can now magically move through windows

**Why is this bad/What are the consequences**:

Shouldn't be able to teleport through a window like that

**Steps to reproduce the problem**:

Be in a place like medbay with the opening lobby, peach on the medical records / crew monitor computer, going through glass

### PR DESCRIPTION
**Pull request by dearmochi**
## What Does This PR Do

- **Parrot stolen item now renders under the parrot's icon to make it more clear when it holds something - remember to harm the parrot to make it drop the item**
- **Parrot adjacency checks now take windows into account - this fixes issues where the parrot would teleport behind a window to perch/steal an item (fixes #13640)**
- **Parrots no longer try to perch/steal an item in a turf that is inaccessible to them infinitely, causing them to be stuck**
- Adds emotes when parrot drops item
- Changes default parrot name to not be improper
- Cleans up code where possible
- Corrects grammar errors in parrot messages
- Fixes attacking parrots in harm mode not triggering specific behaviour

## Why It's Good For The Game

~Parrots are an integral part of SS13 and I couldn't imagine this game without them~

- Makes it more clear when a parrot stole something (I tried to make the held item's icon float as well but couldn't)
- Parrots can no longer teleport through walls, just like in real life - better immersion
- Parrots should get stuck less often (or at all) because of path-finding issues that have now been fixed - better immersion
- Fixes grammar in sentences relating to parrots

## Changelog
:cl:
add: Add the icon of the object under the parrot's icon when it steals something
fix: Parrots can no longer teleport through windows
fix: Fix some (possibly all) cases of parrots getting stuck in a loop trying to reach a place
fix: Fix harming parrots not triggering their specific behaviour
spellcheck: Fix parrot grammar
/:cl:

### CODE REVIEW

File: code/modules/mob/living/simple_animal/parrot.dm [hunk 1]
```diff
@@ -115,9 +115,11 @@
 /mob/living/simple_animal/parrot/death(gibbed)
 	if(can_die())
 		if(held_item)
-			held_item.loc = src.loc
+			custom_emote(EMOTE_VISUAL, "lets go of [held_item]!")
+			held_item.loc = loc
```
Comment (Fox-McCloud) [inline code/modules/mob/living/simple_animal/parrot.dm:119]: forceMove

File: code/modules/mob/living/simple_animal/parrot.dm [hunk 2]
```diff
@@ -645,16 +670,18 @@
 	if(!drop_gently)
 		if(istype(held_item, /obj/item/grenade))
 			var/obj/item/grenade/G = held_item
-			G.loc = src.loc
+			G.loc = loc
```
Comment (Fox-McCloud) [inline code/modules/mob/living/simple_animal/parrot.dm:673]: forceMove

File: code/modules/mob/living/simple_animal/parrot.dm [hunk 3]
```diff
@@ -645,16 +670,18 @@
...
+			to_chat(src, "You let go of [held_item]!")
 			held_item = null
+			update_held_icon()
 			return 1
 
-	to_chat(src, "You drop the [held_item].")
+	to_chat(src, "You drop [held_item].")
 
-	held_item.loc = src.loc
+	held_item.loc = loc
```
Comment (Fox-McCloud) [inline code/modules/mob/living/simple_animal/parrot.dm:682]: and so on.

File: code/modules/mob/living/simple_animal/parrot.dm [hunk 4]
```diff
@@ -666,10 +693,10 @@
 		return
 
 	if(icon_state == "parrot_fly")
-		for(var/atom/movable/AM in view(src,1))
+		for(var/atom/movable/AM in view(src, 1))
 			for(var/perch_path in desired_perches)
 				if(istype(AM, perch_path))
-					src.loc = AM.loc
+					loc = AM.loc
```
Comment (Fox-McCloud) [inline code/modules/mob/living/simple_animal/parrot.dm:699]: Same.

File: code/game/objects/objs.dm [hunk 5]
```diff
@@ -314,8 +314,14 @@ a {
 /obj/proc/container_resist(var/mob/living)
 	return
 
-/obj/proc/CanAStarPass()
+/obj/proc/CanAStarPass(ID, dir, caller)
 	. = !density
+	if(!. && istype(caller, /mob/living/simple_animal/parrot)) // Only check below if we can't pass to save up on calls
```
Comment (Fox-McCloud) [inline code/game/objects/objs.dm:319]: This is pretty gross and reallyyy snowflakey as far as code goes.

Comment (dearmochi) [reply code/game/objects/objs.dm:319]: I agree but this part is necessary to make AStar consider perches as valid nodes. What if I flipped that and made `/atom/proc/CanAStarPassTo(ID, dir, destination) ` (could even be `mob`) that returns whether the source should consider `destination` as a good node? Then when checking the links during AStar I could make an or that checks if either `CanAStarPass` or `CanAStarPassTo` return `TRUE` then the object is a valid node.

File: code/modules/mob/living/simple_animal/parrot.dm [hunk 6]
```diff
@@ -389,64 +392,75 @@
...
-
+		if(Adjacent(parrot_interest))
 			if(isliving(parrot_interest))
 				steal_from_mob()
-
 			else //This should ensure that we only grab the item we want, and make sure it's not already collected on our perch
 				if(!parrot_perch || parrot_interest.loc != parrot_perch.loc)
 					held_item = parrot_interest
+					update_held_icon()
 					parrot_interest.loc = src
```
Comment (Fox-McCloud) [inline code/modules/mob/living/simple_animal/parrot.dm:408]: forceMove

File: code/modules/mob/living/simple_animal/parrot.dm [hunk 7]
```diff
@@ -389,64 +392,75 @@
...
+
 		if(!parrot_perch || !isturf(parrot_perch.loc)) //Make sure the perch exists and somehow isnt inside of something else.
 			parrot_perch = null
 			parrot_state = PARROT_WANDER
 			return
 
-		if(in_range(src, parrot_perch))
-			src.loc = parrot_perch.loc
+		if(Adjacent(parrot_perch))
+			loc = parrot_perch.loc
```
Comment (Fox-McCloud) [inline code/modules/mob/living/simple_animal/parrot.dm:434]: forceMove

File: code/modules/mob/living/simple_animal/parrot.dm [hunk 8]
```diff
@@ -568,20 +592,20 @@
...
 		//Make sure we're not already holding it and it's small enough
 		if(I.loc != src && I.w_class <= WEIGHT_CLASS_SMALL)
-
 			//If we have a perch and the item is sitting on it, continue
 			if(!client && parrot_perch && I.loc == parrot_perch.loc)
 				continue
 
 			held_item = I
+			update_held_icon()
 			I.loc = src
```
Comment (Fox-McCloud) [inline code/modules/mob/living/simple_animal/parrot.dm:607]: forceMove

Comment (dearmochi) [reply code/modules/mob/living/simple_animal/parrot.dm]: All done in the recent commit.

File: code/_DATASTRUCTURES/heap.dm [hunk 9]
```diff
@@ -3,28 +3,27 @@
 //Heap object
 //////////////////////
 
-/Heap
+/datum/Heap
```
Comment (AffectedArc07) [inline code/_DATASTRUCTURES/heap.dm:6]: datumise ALL the things

File: code/modules/mob/living/simple_animal/parrot.dm [hunk 10]
```diff
@@ -611,8 +634,9 @@
 		if(stolen_item)
 			C.unEquip(stolen_item)
```
Comment (farie82) [inline code/modules/mob/living/simple_animal/parrot.dm:635]: No check on this if it actually succeeds or not.
Nodrop items cannot be taken from somebody and those will break here

File: code/game/objects/items.dm [hunk 11]
```diff
@@ -19,7 +19,6 @@ GLOBAL_DATUM_INIT(fire_overlay, /image, image("icon" = 'icons/goonstation/effect
 	can_be_hit = FALSE
 	suicidal_hands = TRUE
 
-	var/list/attack_verb //Used in attackby() to say how something was attacked "[x] has been [z.attack_verb] by [y] with [z]"
```
[the remark this answers is not in the document]
Comment (farie82) [reply code/game/objects/items.dm:22]: Still here. Same with the stuff in objs.dm

[the remark this answers is not in the document]
Comment (dearmochi) [reply code/game/objects/items.dm:22]: sorry, I got confused for a moment. Fixed now

File: code/game/objects/items.dm [hunk 12]
```diff
@@ -112,6 +113,8 @@ GLOBAL_DATUM_INIT(fire_overlay, /image, image("icon" = 'icons/goonstation/effect
 			hitsound = 'sound/items/welder.ogg'
 		if(damtype == "brute")
 			hitsound = "swing_hit"
+	if(!attack_verb)
```
Comment (AffectedArc07) [inline code/game/objects/items.dm:116]: `LAZYINITLIST(attack_verb)`

```dm
#define LAZYINITLIST(L) if (!L) L = list()
```

File: code/game/objects/objs.dm [hunk 13]
```diff
@@ -307,12 +307,12 @@ a {
...
 	return FALSE
 
-/obj/water_act(volume, temperature, source, method = REAGENT_TOUCH)
+/obj/water_act(volume = 0, temperature = 0, source, method = REAGENT_TOUCH)
 	. = ..()
 	extinguish()
 	acid_level = 0
 
-/obj/singularity_pull(S, current_size)
+/obj/singularity_pull(S, current_size = 0)
```
Comment (farie82) [inline code/game/objects/objs.dm:315]: The default here does not make much sense. As the singulo should always pass it's size

File: code/game/objects/items.dm (hunk 12 above)
Comment (dearmochi) [reply code/game/objects/items.dm:116]: done

File: code/game/objects/objs.dm [hunk 14]
```diff
@@ -307,12 +307,12 @@ a {
 		return TRUE
 	return FALSE
 
-/obj/water_act(volume, temperature, source, method = REAGENT_TOUCH)
+/obj/water_act(volume = 0, temperature = 0, source, method = REAGENT_TOUCH)
```
Comment (farie82) [inline code/game/objects/objs.dm:310]: These having defaults doesn't really fit I think

Comment (dearmochi) [reply code/game/objects/objs.dm]: that was just me being overzealous. Fixed

**Review by AffectedArc07 — APPROVED** [review]

**Pull request merged**